File: java/src/main/java/com/sketches_oxide/frequency/CountSketch.java
package com.sketches_oxide.frequency;

import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * CountSketch - Unbiased Frequency Estimation.
 *
 * <p>CountSketch provides unbiased frequency estimation with bounded error.
 * Unlike Count-Min Sketch which only overestimates, CountSketch can both
 * over- and under-estimate but provides an unbiased estimate on average.
 *
 * <p><strong>Algorithm Overview:</strong>
 * <p>CountSketch uses multiple hash functions with random sign flips (+1/-1).
 * Each item is hashed to multiple positions, and a random sign is applied
 * before incrementing. The estimate is the median of signed counts, which
 * cancels out collisions in expectation.
 *
 * <p><strong>Key Differences from Count-Min Sketch:</strong>
 * <ul>
 *   <li>Unbiased estimates (can under-estimate or over-estimate)</li>
 *   <li>Uses median instead of minimum for estimation</li>
 *   <li>Better for scenarios where unbiased estimates are preferred</li>
 *   <li>Supports negative frequency updates</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Heavy hitter detection with unbiased estimates</li>
 *   <li>Network traffic monitoring</li>
 *   <li>Natural language processing word frequencies</li>
 *   <li>Scenarios requiring unbiased frequency estimation</li>
 * </ul>
 *
 * <p><strong>Space Complexity:</strong> O(1/epsilon^2 * ln(1/delta))
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (CountSketch cs = new CountSketch(0.01, 0.01)) {
 *     cs.update("apple");
 *     cs.update("apple");
 *     cs.update("banana");
 *
 *     // Estimate is unbiased (may be slightly above or below true count)
 *     System.out.println("Estimated apple count: " + cs.estimate("apple"));
 * }
 * </pre>
 *
 * @see CountMinSketch
 */
public final class CountSketch extends NativeSketch implements MergeableSketch<CountSketch> {

    private final double epsilon;
    private final double delta;

    /**
     * Create a new CountSketch.
     *
     * @param epsilon relative error bound (0 < epsilon < 1)
     *                Smaller values mean more accurate but larger space
     * @param delta failure probability (0 < delta < 1)
     *              Smaller values mean more confident but larger space
     * @throws IllegalArgumentException if parameters are invalid
     * @throws OutOfMemoryException if native memory allocation fails
     */
    public CountSketch(double epsilon, double delta) {
        if (epsilon <= 0 || epsilon >= 1) {
            throw new IllegalArgumentException("epsilon must be in range (0, 1), got: " + epsilon);
        }
        if (delta <= 0 || delta >= 1) {
            throw new IllegalArgumentException("delta must be in range (0, 1), got: " + delta);
        }

        this.epsilon = epsilon;
        this.delta = delta;
        this.nativePtr = SketchOxideNative.countsketch_new(epsilon, delta);

        if (this.nativePtr == 0) {
            throw new OutOfMemoryException("Failed to allocate CountSketch");
        }
    }

    /**
     * Update the frequency of an item (increment by 1).
     *
     * @param item the item to count
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.countsketch_update(nativePtr, item);
    }

    /**
     * Update with a string item.
     *
     * @param item the string to count
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        update(item.getBytes());
    }

    /**
     * Estimate the frequency of an item.
     *
     * <p>Returns an unbiased estimate that may be above or below the true count.
     * The estimate is the median of signed counts across all hash functions.
     *
     * @param item the item to query
     * @return estimated frequency (may be negative due to hash collisions)
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public long estimate(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.countsketch_estimate(nativePtr, item);
    }

    /**
     * Estimate the frequency of a string item.
     *
     * @param item the string to query
     * @return estimated frequency
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public long estimate(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return estimate(item.getBytes());
    }

    /**
     * Merge another CountSketch into this one.
     *
     * <p>Both sketches must have the same epsilon and delta parameters.
     *
     * @param other the sketch to merge
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if parameters don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(CountSketch other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.epsilon != other.epsilon || this.delta != other.delta) {
            throw new IllegalArgumentException(
                    "Cannot merge CountSketches with different parameters: " +
                            String.format("(%.3f, %.3f) vs (%.3f, %.3f)",
                                    this.epsilon, this.delta, other.epsilon, other.delta));
        }

        SketchOxideNative.countsketch_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Get the width parameter (number of counters per row).
     *
     * @return the width
     * @throws IllegalStateException if the sketch has been closed
     */
    public int width() {
        checkAlive();
        return SketchOxideNative.countsketch_width(nativePtr);
    }

    /**
     * Get the depth parameter (number of hash functions/rows).
     *
     * @return the depth
     * @throws IllegalStateException if the sketch has been closed
     */
    public int depth() {
        checkAlive();
        return SketchOxideNative.countsketch_depth(nativePtr);
    }

    /**
     * Get the epsilon parameter.
     *
     * @return the epsilon (relative error bound)
     */
    public double epsilon() {
        return epsilon;
    }

    /**
     * Get the delta parameter.
     *
     * @return the delta (failure probability)
     */
    public double delta() {
        return delta;
    }

    /**
     * Serialize the sketch to binary format.
     *
     * @return binary representation of the sketch
     * @throws IllegalStateException if the sketch has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.countsketch_serialize(nativePtr);
    }

    /**
     * Deserialize from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new CountSketch instance
     * @throws IllegalArgumentException if data is invalid
     * @throws NullPointerException if data is null
     */
    public static CountSketch deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.countsketch_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized CountSketch data");
        }

        CountSketch cs = new CountSketch(0.01, 0.01);
        SketchOxideNative.countsketch_free(cs.nativePtr);
        cs.nativePtr = ptr;
        return cs;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.countsketch_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        try {
            if (nativePtr != 0) {
                return String.format("CountSketch(epsilon=%.3f, delta=%.3f, width=%d, depth=%d)",
                        epsilon, delta, width(), depth());
            }
        } catch (IllegalStateException e) {
            // Closed
        }
        return "CountSketch(closed)";
    }
}

File: dotnet/SketchOxide/src/Frequency/CountSketch.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Frequency;

/// <summary>
/// Count Sketch for frequency estimation with support for positive and negative updates.
///
/// Unlike Count-Min Sketch, Count Sketch can provide unbiased frequency estimates
/// and can handle both increments and decrements. The sketch uses width and depth
/// parameters to control accuracy and memory usage.
/// </summary>
public sealed class CountSketch : NativeSketch, IMergeableSketch<CountSketch>
{
    private readonly uint _width;
    private readonly uint _depth;

    /// <summary>
    /// Creates a new Count Sketch with the specified dimensions.
    /// </summary>
    /// <param name="width">Width of the sketch (number of counters per row). Must be greater than 0.</param>
    /// <param name="depth">Depth of the sketch (number of hash functions/rows). Must be greater than 0.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if width or depth is 0.</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public CountSketch(uint width, uint depth)
    {
        if (width == 0)
            throw new ArgumentOutOfRangeException(nameof(width), width, "Width must be greater than 0");
        if (depth == 0)
            throw new ArgumentOutOfRangeException(nameof(depth), depth, "Depth must be greater than 0");

        _width = width;
        _depth = depth;
        NativePtr = SketchOxideNative.countsketch_new(width, depth);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native CountSketch");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private CountSketch(uint width, uint depth, nuint ptr)
    {
        _width = width;
        _depth = depth;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the width (number of counters per row) of this sketch.
    /// </summary>
    public uint Width
    {
        get
        {
            CheckAlive();
            return _width;
        }
    }

    /// <summary>
    /// Gets the depth (number of hash functions) of this sketch.
    /// </summary>
    public uint Depth
    {
        get
        {
            CheckAlive();
            return _depth;
        }
    }

    /// <summary>
    /// Updates the sketch with a new element, incrementing its count by 1.
    /// </summary>
    /// <param name="data">The bytes representing the item to count.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.countsketch_update(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Updates the sketch with a string element, incrementing its count by 1.
    /// </summary>
    /// <param name="value">The string item to count.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Update(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Estimates the frequency count of an element.
    /// </summary>
    /// <param name="data">The bytes representing the item to query.</param>
    /// <returns>The estimated count for the item. Can be negative due to the nature of Count Sketch.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Estimate(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.countsketch_estimate(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Estimates the frequency count of a string element.
    /// </summary>
    /// <param name="value">The string item to query.</param>
    /// <returns>The estimated count for the item. Can be negative due to the nature of Count Sketch.</returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Estimate(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Estimate(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Merges another Count Sketch into this one.
    /// </summary>
    /// <param name="other">The sketch to merge. Must have the same width and depth.</param>
    /// <exception cref="ArgumentNullException">Thrown if other is null.</exception>
    /// <exception cref="ArgumentException">Thrown if sketches have different dimensions.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if either sketch is disposed.</exception>
    public void Merge(CountSketch other)
    {
        CheckAlive();
        if (other == null) throw new ArgumentNullException(nameof(other));
        other.CheckAlive();

        if (_width != other._width || _depth != other._depth)
            throw new ArgumentException($"Cannot merge sketches with different dimensions: ({_width}, {_depth}) != ({other._width}, {other._depth})");

        SketchOxideNative.countsketch_merge(NativePtr, other.NativePtr);
    }

    /// <summary>
    /// Serializes the sketch to a byte array.
    /// </summary>
    /// <returns>Serialized sketch bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.countsketch_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a Count Sketch from a byte array.
    /// </summary>
    /// <param name="data">Serialized sketch bytes.</param>
    /// <param name="width">The width parameter used when creating the original sketch.</param>
    /// <param name="depth">The depth parameter used when creating the original sketch.</param>
    /// <returns>A new CountSketch instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static CountSketch Deserialize(byte[] data, uint width, uint depth)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.countsketch_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize CountSketch: invalid data");

        return new CountSketch(width, depth, ptr);
    }

    /// <summary>
    /// Returns a string representation of the sketch.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "CountSketch(disposed)";
        return $"CountSketch(width={_width}, depth={_depth})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.countsketch_free(NativePtr);
            NativePtr = 0;
        }
    }
}
